sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp

```diff
@@ -0,0 +1,452 @@
+  }
+}
+
+template <typename Tm, typename Tc, std::size_t M, std::size_t K, std::size_t N,
+          sycl::ext::oneapi::experimental::matrix::layout LayoutA,
+          sycl::ext::oneapi::experimental::matrix::layout LayoutB>
+void joint_matrix_mad_hip(
+    joint_matrix_hip<
+        Tc, sycl::ext::oneapi::experimental::matrix::use::accumulator, M, N,
+        sycl::ext::oneapi::experimental::matrix::layout::dynamic> &D,
+    const joint_matrix_hip<Tm, sycl::ext::oneapi::experimental::matrix::use::a,
+                           M, K, LayoutA> &A,
+    const joint_matrix_hip<Tm, sycl::ext::oneapi::experimental::matrix::use::b,
+                           K, N, LayoutB> &B,
+    const joint_matrix_hip<
+        Tc, sycl::ext::oneapi::experimental::matrix::use::accumulator, M, N,
+        sycl::ext::oneapi::experimental::matrix::layout::dynamic> &C) {
+  if constexpr (std::is_same_v<Tm, sycl::half>) {
+    if constexpr (M == 16 && N == 16) {
+      D.data = __builtin_amdgcn_mfma_f32_16x16x16f16(A.data, B.data, C.data, 0,
+                                                     0, 0);
+    } else if constexpr (M == 32 && N == 32) {
+      D.data =
+          __builtin_amdgcn_mfma_f32_32x32x8f16(A.data, B.data, C.data, 0, 0, 0);
+    }
+  } else if constexpr (std::is_same_v<Tm, bfloat16>) {
+    if constexpr (M == 16 && N == 16) {
+      D.data = __builtin_amdgcn_mfma_f32_16x16x16bf16_1k(A.data, B.data, C.data,
+                                                         0, 0, 0);
+    } else if constexpr (M == 32 && N == 32) {
+      D.data = __builtin_amdgcn_mfma_f32_32x32x8bf16_1k(A.data, B.data, C.data,
+                                                        0, 0, 0);
+    }
+  } else if constexpr (std::is_same_v<Tm, double>) {
+    if constexpr (M == 16 && N == 16) {
+      D.data = __builtin_amdgcn_mfma_f64_16x16x4f64(A.data[0], B.data[0],
+                                                    C.data, 0, 0, 0);
+    }
+  } else if constexpr (std::is_same_v<Tm, int8_t>) {
+    if constexpr (M == 16 && N == 16) {
+      D.data = __builtin_amdgcn_mfma_i32_16x16x16i8(
+          *reinterpret_cast<const Tc *>(A.data),
+          *reinterpret_cast<const Tc *>(B.data), C.data, 0, 0, 0);
+    } else if constexpr (M == 32 && N == 32) {
+      D.data = __builtin_amdgcn_mfma_i32_32x32x8i8(
+          *reinterpret_cast<const Tc *>(A.data),
+          *reinterpret_cast<const Tc *>(B.data), C.data, 0, 0, 0);
+    }
+  } else {
+    static_assert(false && "Invalid configuration!");
+  }
+}
+
+template <typename S, size_t M, size_t N,
+          sycl::ext::oneapi::experimental::matrix::use Use,
+          sycl::ext::oneapi::experimental::matrix::layout Layout, typename F>
+void joint_matrix_apply(joint_matrix_hip<S, Use, M, N, Layout> &jm,
+                        F &&lambda) {
+  if constexpr (std::is_same_v<S, double> &&
+                Use !=
+                    sycl::ext::oneapi::experimental::matrix::use::accumulator) {
+    jm.data[0] = lambda(jm.data[0]);
```

Comment:
nit: I see what you've done here. I'm not sure if in the future this is guaranteed to work as more cases are added (although it might). It might be worth at least considering using a marray for the jm.data, which simplifies the implementation of `joint_matrix_apply`, (see cuda impl).